container/common/src/main/java/org/mobicents/slee/container/management/jmx/SleeManagementMBeanImpl.java

```diff
@@ -220,19 +237,134 @@ private void notifyStateChange(SleeState oldState, SleeState newState) {
 						newState, oldState, sleeStateChangeSequenceNumber++));
 	}
 
+	public static Logger getLogger() {
+		return logger;
+	}
+
+	public static void setLogger(Logger logger) {
+		SleeManagementMBeanImpl.logger = logger;
+	}
+
+	public int getDefaultActiveSessionsThreshold() {
+		return defaultActiveSessionsThreshold;
+	}
+
+	public void setDefaultActiveSessionsThreshold(int defaultActiveSessionsThreshold) {
+		this.defaultActiveSessionsThreshold = defaultActiveSessionsThreshold;
+		if(this.sleeContainer != null) {
+			this.sleeContainer.setGracefulStopActivitiesCountThreshold(defaultActiveSessionsThreshold);
+		}
+	}
+
+	public long getDefaultGracefulShutdownWaitTime() {
+		return defaultGracefulShutdownWaitTime;
+	}
+
+	public void setDefaultGracefulShutdownWaitTime(long defaultGracefulShutdownWaitTime) {
+		this.defaultGracefulShutdownWaitTime = defaultGracefulShutdownWaitTime;
+		if(this.sleeContainer != null) {
+			this.sleeContainer.setGracefulStopWaitTime(defaultGracefulShutdownWaitTime);
+		}
+	}
+
 	/**
 	 * Gracefully stop the SLEE. Should do it in a non-blocking manner.
+	 * Requests SLEE container to enter graceful STOPPING state.
+	 * @return Status and the number of active sessions remaining after invocation of Graceful Shutdown command
+	 */
+	public String gracefulStop(Integer ast, Long time) throws InvalidStateException, ManagementException {
+		if(ast != null && ast >= 0){
+			sleeContainer.setGracefulStopActivitiesCountThreshold(ast);
+		} else {
+			logger.info("Using default defaultActiveSessionsThreshold value for graceful shutdown (" + defaultActiveSessionsThreshold + ")" );
+			sleeContainer.setGracefulStopActivitiesCountThreshold(defaultActiveSessionsThreshold);
+		}
+		if(time != null && time > 0) {
+			sleeContainer.setGracefulStopWaitTime(time);
+		} else {
+			logger.info("Using default defaultGracefulShutdownWaitTime value for graceful shutdown (" + defaultGracefulShutdownWaitTime + ")");
+			sleeContainer.setGracefulStopWaitTime(defaultGracefulShutdownWaitTime);
+		}
+
+		logger.info("gracefulStop called. ast=" + sleeContainer.getGracefulStopActivitiesCountThreshold() + " time=" + sleeContainer.getGracefulStopWaitTime());
+
+		stop(false, true);
+
+		return "Container Graceful Shutdown requested successfully.";
+	}
+
+	/**
+	 * Stop the SLEE. Should do it in a non-blocking manner.
 	 * 
 	 * @see javax.slee.management.SleeManagementMBean#stop()
 	 */
 	public void stop() throws InvalidStateException, ManagementException {

```

Comment:
why do we need a nonGraceful stop? As stated in Spec all stops should be gracefull, entering the STOPPING state, and causing all RAs to enter stopping state as well...14.5.1 Operational states of the SLEE